src/lazy.rs

```diff
@@ -54,3 +58,57 @@ impl LazyBool {
         self.0.unsync_init(|| init() as usize) != 0
     }
 }
+
+// This structure represents a lazily initialized static pointer value.
+///
+/// It's intended to be used for weak linking of a C function that may
+/// or may not be present at runtime.
+///
+/// Based off of the DlsymWeak struct in libstd:
+/// https://github.com/rust-lang/rust/blob/1.61.0/library/std/src/sys/unix/weak.rs#L84
+/// except that the caller must manually cast self.ptr() to a function pointer.
+pub struct LazyPtr {
+    addr: AtomicPtr<c_void>,
+}
+
+impl LazyPtr {
+    /// A non-null pointer value which indicates we are uninitialized.
+    ///
+    /// This constant should ideally not be a valid pointer.
+    /// However, if by chance initialization function passed to the `get`
+    /// method does return UNINIT, there will not be undefined behavior.
+    /// The initialization function will just be called each time `get()`
+    /// is called. This would be inefficient, but correct.
+    const UNINIT: *mut c_void = !0usize as *mut c_void;
+
+    /// Construct new `LazyPtr` in uninitialized state.
+    pub const fn new() -> Self {
+        Self {
+            addr: AtomicPtr::new(Self::UNINIT),
+        }
+    }
+
+    // Runs the init() function at most once, returning the value of some run of
+    // init(). Multiple callers can run their init() functions in parallel.
+    // init() should always return the same value, if it succeeds.
+    pub fn unsync_init(&self, f: impl Fn() -> *mut c_void) -> *mut c_void {
+        // Despite having only a single atomic variable (self.addr), we still
+        // cannot always use Ordering::Relaxed, as we need to make sure a
+        // successful call to `f` is "ordered before" any data read through
+        // the returned pointer (which occurs when the function is called).
+        // Our implementation mirrors that of the one in libstd, meaning that
+        // the use of non-Relaxed operations is probably unnecessary.
+        match self.addr.load(Ordering::Relaxed) {
```

Comment:
Maybe we should use this code instead?
```rust
match self.addr.load(Ordering::Acquire) {
    Self::UNINIT => {
        let addr = f();
        self.addr.store(addr, Ordering::Release);
        addr
    }
    addr => addr,
}
```
I don't think "relaxed load + acquire fence" has advantages of simply doing "acquire load" in this case.  If anything, it may be even worse.

Reply:
Yeah I think doing the simpler acquire/release, is more obviously correct. I don't think we really gain anything in practice from using a relaxed load.